arangod/GeoIndex/Covering.h

```diff
@@ -0,0 +1,124 @@
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <s2/s2cap.h>
+#include <s2/s2cell_id.h>
+#include <s2/s2region.h>
+#include <s2/s2region_coverer.h>
+
+#include <queue>
+#include <vector>
+
+#include "Geo/GeoParams.h"
+#include "Geo/Utils.h"
+#include "VocBase/Identifiers/LocalDocumentId.h"
+
+namespace arangodb {
+namespace geo_index {
+
+/// @brief Helper class to build a simple covering query iterator.
+/// Will return findings not sorted by anything, will filter according
+/// to the filterType. Should be storage engine agnostic.
+class CoveringUtils {
+ public:
+  /// @brief Type of documents buffer
+  typedef std::deque<LocalDocumentId> GeoDocumentsQueue;
+
+  explicit CoveringUtils(geo::QueryParams&& params) noexcept;
+  ~CoveringUtils();
+
+ public:
+  inline geo::FilterType filterType() const { return _params.filterType; }
+
+  inline geo::ShapeContainer const& filterShape() const {
+    return _params.filterShape;
+  }
+
+  /// @brief all intervals are covered, no more buffered results
+  bool isDone() const { return _buffer.empty() && _allIntervalsCovered; }
+
+  /// @brief has buffered results
+  inline bool hasNext() const { return !_buffer.empty(); }
+
+  inline size_t bufferSize() const { return _buffer.size(); }
+
+  LocalDocumentId const& getNext() const {
+    TRI_ASSERT(hasNext());
+    return _buffer.front();
+  }
+
+  inline void next() {
+    TRI_ASSERT(hasNext());
+    _buffer.pop_front();
+  }
+
+  /// @brief reset query to initial state
+  void reset();
+
+  /// Call only when current scan intervals contain no more results
+  std::vector<geo::Interval> intervals();
+
+  /// buffer and sort results
+  void reportFound(LocalDocumentId lid, S2Point const& center);
+
+  /// @brief Reference to parameters used
+  geo::QueryParams const& params() const { return _params; }
+
+  size_t _found = 0;
+  size_t _rejection = 0;
+
+ private:
+  inline bool isFilterNone() const noexcept {
+    return _params.filterType == geo::FilterType::NONE;
+  }
+
+  inline bool isFilterContains() const noexcept {
+    return _params.filterType == geo::FilterType::CONTAINS;
+  }
+
+  inline bool isFilterIntersects() const noexcept {
+    return _params.filterType == geo::FilterType::INTERSECTS;
+  }
+
+ private:
+  geo::QueryParams const _params;
+
+  bool _allIntervalsCovered;
+
+  /// Total number of interval calculations
+  size_t _numScans;
+
+  /// buffer of found documents
+  GeoDocumentsQueue _buffer;
+
+  // deduplication filter
+  std::unordered_set<uint64_t> _seenDocs;
```

Comment:
What about using `containers::FlatHashSet` or `containers::NodeHashSet` if you really need pointer stability?

Reply:
We have agreed to do concentrate on correctness for this PR, it was difficult enough to achieve that. We will do performance improvements in a separate PR. I will put it on the list of possible performance improvements.